frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -83,36 +223,73 @@ self.addEventListener('message', function (event) {
 // Handle clicks on notifications issued via registration.showNotification().
 self.addEventListener('notificationclick', event => {
   console.debug('[sw] Notification clicked:', event.notification)
-})
+  event.notification.close()
 
-self.addEventListener('push', function (event) {
-  // PushEvent reference: https://developer.mozilla.org/en-US/docs/Web/API/PushEvent
+  event.waitUntil(
+    self.clients.matchAll({ type: 'window' }).then((clientList) => {
+      clientList.sort((a, b) => {
+        if (a.focused !== b.focused) {
+          return a.focused ? -1 : 1
+        }
+        if (a.visibilityState !== b.visibilityState) {
+        // order is visible, prerender, hidden
+          if (a.visibilityState === 'visible') return -1
+          if (b.visibilityState === 'visible') return 1
+          if (a.visibilityState === 'hidden') return 1
+          if (b.visibilityState === 'hidden') return -1
+        }
 
-  if (!(self.Notification && self.Notification.permission === 'granted')) {
-    console.debug("[sw] received a push notification but aren't displaying it due to the permission not granted")
-    return
-  }
-
-  const data = event.data.json()
-  console.debug('[sw] push received: ', data)
+        return 0
+      })
+      if (!clientList.length) {
+        return self.clients.openWindow(`/app${event.notification.data.path ?? '/'}`)
+      }
+      const client = clientList[0]
+      if (event.notification.data?.path) {
+        client.postMessage({
+          type: 'navigate',
+          path: event.notification.data.path
+        })
+      }
+      if (!client.focused) return client.focus()
+    }))
+})
 
-  self.registration.showNotification(
-    data.title,
-    {
-      body: data.body || '',
-      icon: '/assets/images/pwa-icons/group-income-icon-transparent.svg'
+sbp('okTurtles.events/on', KV_EVENT, ({ contractID, key, data }) => {
+  const rootState = sbp('chelonia/rootState')
+  switch (key) {
+    case KV_KEYS.LAST_LOGGED_IN: {
+      if (!rootState.lastLoggedIn) rootState.lastLoggedIn = Object.create(null)
+      rootState.lastLoggedIn[contractID] = data
```

Comment:
How/where/when do these updates get saved to IndexedDB?

Reply:
They don't. Updated.